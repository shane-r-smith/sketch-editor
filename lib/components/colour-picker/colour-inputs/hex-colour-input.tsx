import { useCallback, useEffect, useState } from "react";
import type { BaseColourPickerProps } from "../colour-picker.types";
import { colord } from "colord";
import TextField from "@mui/material/TextField";
import FormLabel from "@mui/material/FormLabel";
import Stack from "@mui/material/Stack";

export function HexColourInput({
  type,
  colour,
  onChange,
  ...rest
}: BaseColourPickerProps) {
  const [currentColour, setCurrentColour] = useState(colord(colour).toHex());

  const handleChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      setCurrentColour(event.target.value);
      onChange(event.target.value, type);
    },
    [onChange, type]
  );

  useEffect(() => {
    if (currentColour === colour) {
      return;
    }

    const value = colord(colour);

    if (value.isValid()) {
      setCurrentColour(value.toHex());
      return;
    }

    setCurrentColour(colour);
    return;
  }, [colour, currentColour]);

  return (
    <Stack>
      <FormLabel>HEX</FormLabel>
      <TextField
        {...rest}
        value={currentColour}
        onChange={handleChange}
        size="small"
      />
    </Stack>
  );
}
