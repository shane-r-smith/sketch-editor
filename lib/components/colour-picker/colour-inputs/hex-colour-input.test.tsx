import { fireEvent, render } from "@testing-library/react";
import { HexColourInput } from "./hex-colour-input";
import type { BaseColourPickerProps } from "../colour-picker.types";

describe("<HexColourInput />", () => {
  const defaultProps: BaseColourPickerProps = {
    type: "HEX",
    colour: "#ff0000",
    onChange: vi.fn(),
  };

  it("renders correctly and matches snapshot", () => {
    const { container } = render(<HexColourInput {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("calls onChange with updated value and type when input changes", async () => {
    const { container } = render(<HexColourInput {...defaultProps} />);

    const input = container.querySelector("input")!;

    fireEvent.change(input, { target: { value: "#00ff00" } });
    expect(defaultProps.onChange).toHaveBeenCalledWith("#00ff00", "HEX");
  });

  it("handles invalid colour input gracefully", async () => {
    const { container } = render(<HexColourInput {...defaultProps} />);

    const input = container.querySelector("input")!;
    fireEvent.change(input, { target: { value: "invalid" } });

    expect(defaultProps.onChange).toHaveBeenCalledWith("invalid", "HEX");
  });
});
