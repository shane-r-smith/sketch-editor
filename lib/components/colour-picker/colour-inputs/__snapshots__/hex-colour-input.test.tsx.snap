// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<HexColourInput /> > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="MuiStack-root css-nen11g-MuiStack-root"
  >
    <label
      class="MuiFormLabel-root MuiFormLabel-colorPrimary css-nwsdb5-MuiFormLabel-root"
    >
      HEX
    </label>
    <div
      class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
          id="«r0»"
          type="text"
          value="#ff0000"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1nf2c5d-MuiNotchedOutlined-root"
          >
            <span
              aria-hidden="true"
              class="notranslate"
            >
              ​
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
